include/stats/histogram.h

```diff
@@ -0,0 +1,122 @@
+//
+// Copyright (c) 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
+//
+
+#ifndef NCCL_OFI_STATS_HISTOGRAM
+#define NCCL_OFI_STATS_HISTOGRAM
+
+#include <cassert>
+#include <chrono>
+#include <cstddef>
+#include <string>
+#include <sstream>
+#include <vector>
+
+#include "nccl_ofi_log.h"
+#include "histogram_binner.h"
+
+
+//
+// Base histogram class.  Histograms are a lightweight mechanism for tracking
+// events occurances in code and are used for instrumenting the plugin code.
+//
+// T is the type of the data that will be inserted into the histogram.  Any POD
+//  will work, and little effort has been put into making the interface safe for
+//  non-Pods.
+//
+template <typename T, typename Binner>
+class histogram {
+public:
+	histogram(const std::string& description_arg, Binner binner_arg)
+		: description(description_arg), binner(binner_arg),
+		bins(binner.get_num_bins()), num_samples(0), first_insert(true)
+	{
+	}
+
+	void insert(const T& input_val)
+	{
+		if (OFI_UNLIKELY(first_insert)) {
+			max_val = min_val = input_val;
+			first_insert = false;
+		}
```

Comment:
Does this need to be atomic for threads?

Reply:
the interface is not thread safe.  I assume there's locking somewhere else.